# Player Data Kaggle-Fjelstul

## Objectives

Develop a dataset with Kaggle dataset ID's, Fjestul Dataset IDs, and Player Names.

This will serve as a mathcing tool to assign season data (from the kaggle dataset) to world cup data (from the Fjestul dataset)

## Inputs

- `1.DataCleaning-R/Data/RDS/FullRoster.rds`
- `1.DataCleaning-R/kaggle_data/players.csv`

## Output

- `1.DataCleaning-R/Data/RDS/Matching.rds`

## Acknowledgments

Data in this notebook is taken from Kaggle using Kaggle's API. The dataset in Kaggle is made from data originally from Transfermarket.com.

## Libraries

In [2]:
library(dplyr)
library(here)
library(stringr)
library(stringi)



Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union


here() starts at /Users/eialnisman/Desktop/WC2026Forecast



## Player data (Kaggle)

### Call Data with our Kaggle API

In [3]:
system("kaggle datasets download davidcariboo/player-scores -p kaggle_data")

unzip("kaggle_data/player-scores.zip",exdir = "kaggle_data")


Warning message in unzip("kaggle_data/player-scores.zip", exdir = "kaggle_data"):
"zip file is corrupt"


In [4]:
player_data <- read.csv(here("1.DataCleaning-R", "kaggle_data", "players.csv"), stringsAsFactors = FALSE, encoding = "UTF-8")

colnames(player_data)

[1] "player_id"                           
 [2] "first_name"                          
 [3] "last_name"                           
 [4] "name"                                
 [5] "last_season"                         
 [6] "current_club_id"                     
 [7] "player_code"                         
 [8] "country_of_birth"                    
 [9] "city_of_birth"                       
[10] "country_of_citizenship"              
[11] "date_of_birth"                       
[12] "sub_position"                        
[13] "position"                            
[14] "foot"                                
[15] "height_in_cm"                        
[16] "contract_expiration_date"            
[17] "agent_name"                          
[18] "image_url"                           
[19] "international_caps"                  
[20] "international_goals"                 
[21] "current_national_team_id"            
[22] "url"                                 
[23] "current_club_domestic_competition_id"
[24] "current_club_name"                   
[25] "market_value_in_eur"                 
[26] "highest_market_value_in_eur"

In [5]:
player_names <- player_data %>%
    select(player_id, name, last_season)



## Clean Data

### Lets get the player names and IDs from Fjestul

In [6]:
worldcup_players <- readRDS(here("1.DataCleaning-R", "Data", "RDS", "FullRoster.rds"))

worldcup_players <- worldcup_players %>%
    select(player_id, full_name, tournament_id) %>%
        arrange(desc(tournament_id)) %>%
        distinct(player_id, .keep_all = TRUE)


Since the dataset for worldcup_players and player_namess come from sepearate orders their player_id system is different. It is also likely that some names may not match exaclty, lets investigate this possibility. 

In [7]:
length(setdiff(worldcup_players$full_name, player_names$name))
length(worldcup_players$full_name)
setdiff(worldcup_players$full_name, player_names$name)

[1] 856

[1] 2379

[1] "Juli<U+00E1>n <U+00C1>lvarez" "Milo<U+0161> Degenek"      
  [3] "Fran Kara<U+010D>i<U+0107>" "Bailey Wright"             
  [5] "Cameron Devlin"             "not applicable Alisson"    
  [7] "not applicable Danilo"      "not applicable Marquinhos" 
  [9] "not applicable Casemiro"    "not applicable Fred"       
 [11] "not applicable Richarlison" "not applicable Neymar"     
 [13] "not applicable Raphinha"    "not applicable Weverton"   
 [15] "not applicable Fabinho"     "not applicable Antony"     
 [17] "Vin<U+00ED>cius J<U+00FA>nior" "not applicable Rodrygo"    
 [19] "<U+00C9>verton Ribeiro"     "not applicable Ederson"    
 [21] "not applicable Bremer"      "not applicable Pedro"      
 [23] "Jerome Ngom Mbekeli"        "Nicolas Nkoulou"           
 [25] "Moumi Ngamaleu"             "Georges-K<U+00E9>vin Nkoudou"
 [27] "Andr<U+00E9>-Frank Zambo Anguissa" "Eric Maxim Choupo-Moting"  
 [29] "Samuel Gouet"               "Nouhou Tolo"               
 [31] "Souaibou Marou"             "Stephen Eust<U+00E1>quio"  
 [33] "Liam Fraser"                "Lucas Cavallini"           
 [35] "Daniel Chac<U+00F3>n"       "Keysher Fuller"            
 [37] "Johan Venegas"              "Gerson Torres"             
 [39] "Douglas L<U+00F3>pez"       "R<U+00F3>nald Matarrita"   
 [41] "Dominik Livakovi<U+0107>"   "Josip Stani<U+0161>i<U+0107>"
 [43] "Borna Bari<U+0161>i<U+0107>" "Mateo Kova<U+010D>i<U+0107>"
 [45] "Andrej Kramari<U+0107>"     "Ivo Grbi<U+0107>"          
 [47] "Bruno Petkovi<U+0107>"      "Mislav Or<U+0161>i<U+0107>"
 [49] "Jo<U+0161>ko Gvardiol"      "Josip Juranovi<U+0107>"    
 [51] "Ivica Ivu<U+0161>i<U+0107>" "Josip <U+0160>utalo"       
 [53] "Luka Su<U+010D>i<U+0107>"   "Kristijan Jaki<U+0107>"    
 [55] "Joakim M<U+00E6>hle"        "Jens Stryger Larsen"       
 [57] "Frederik R<U+00F8>nnow"     "William Pacho"             
 [59] "Romario Ibarra"             "Michael Estrada"           
 [61] "Xavier Arreaga"             "<U+00C1>ngelo Preciado"    
 [63] "Sebas M<U+00E9>ndez"        "Alexander Dom<U+00ED>nguez"
 [65] "Djorkaeff Reasco"           "Matteo Guendouzi"          
 [67] "Lucas Hernandez"            "Th<U+00E9>o Hernandez"     
 [69] "Lawrence Ati-Zigi"          "Abdul Fatawu Issahaku"     
 [71] "Ibrahim Danlad"             "Abdul Manaf Nurudeen"      
 [73] "Baba Rahman"                "Shojae Khalilzadeh"        
 [75] "Hossein Kanaanizadegan"     "Rouzbeh Cheshmi"           
 [77] "Mehdi Torabi"               "Ahmad Nourollahi"          
 [79] "Hossein Hosseini"           "Abolfazl Jalali"           
 [81] "Ritsu D<U+014D>an"          "Sh<U+016B>ichi Gonda"      
 [83] "Alfredo Talavera"           "N<U+00E9>stor Araujo"      
 [85] "Henry Mart<U+00ED>n"        "Munir Mohamedi"            
 [87] "Ilias Chair"                "Ahmed Reda Tagnaouti"      
 [89] "Bilal El Khannous"          "Badr Benoun"               
 [91] "Yahya Jabrane"              "Wojciech Szcz<U+0119>sny"  
 [93] "<U+0141>ukasz Skorupski"    "Jakub Kami<U+0144>ski"     
 [95] "Sebastian Szyma<U+0144>ski" "Piotr Zieli<U+0144>ski"    
 [97] "Krzysztof Pi<U+0105>tek"    "Przemys<U+0142>aw Frankowski"
 [99] "Micha<U+0142> Sk<U+00F3>ra<U+015B>" "not applicable Pepe"       
[101] "not applicable Vitinha"     "not applicable Ot<U+00E1>vio"
[103] "Saad Al-Sheeb"              "not applicable R<U+00F3>-R<U+00F3>"
[105] "Abdulaziz Hatem"            "Ali Assadalla"             
[107] "Hassan Al-Haydos"           "Musab Kheder"              
[109] "Homam Ahmed"                "Ismaeel Mohammad"          
[111] "Khalid Muneer"              "Salem Al-Hajri"            
[113] "Yousef Hassan"              "Naif Al-Hadhrami"          
[115] "Mostafa Meshaal"            "Mohammed Al-Rubaie"        
[117] "Ali Al-Bulaihi"             "Mohammed Al-Breik"         
[119] "Abdulellah Al-Malki"        "Hassan Al-Tambakti"        
[121] "Abdulrahman Al-Aboud"       "Moussa N'Diaye"            
[123] "<U+00C9>douard Mendy"       "Moustapha Name"         

We can see a very significant amount of players have mismatched names. 

In [8]:
set.seed(2026)

player_names %>%
    slice_sample(n = 20) %>%
        select(name)

name
<chr>
Volodymyr Bilotserkovets
Sam Lidington
Riccardo Montolivo
Dimitri Legbo
Luis Ortiz
Roman Khodunov
Morten Rasmussen
Colin McMenamin
Afonso Sousa


After manually inspecting I see some players are just spelled slightly differently in one dataset vs the other. For example Antar/Anthar Yahiya. Let's trouble shoot other issues first since this might require a lot of manual comparison which will be easier to do with less observations.

In [9]:
player_names %>%
  filter(grepl("Anthar Yahia", name)) %>%
    select(name)

worldcup_players %>%
  filter(grepl("Antar Yahia", full_name)) %>%
    select(full_name)

name
<chr>
Anthar Yahia


full_name
<chr>
Antar Yahia


Some of the player names also differ in the datasets because of accents. For example Ronald Araujo.

In [10]:
player_names %>%
  filter(grepl("Ronald Araujo", name)) %>%
    select(name)

worldcup_players %>%
  filter(grepl("Ronald Ara", full_name)) %>%
    select(full_name)

name
<chr>
Ronald Araujo


full_name
<chr>
Ronald Ara<U+00FA>jo


The accents seem to be a relevant issue across a lot of players. Let's write a function to normalize all text by removing accents. Also making sure to decode all text in the process.

In [11]:
# This was done with significant help from Codex AI
decode_unicode_tags <- function(x) {
  fix_one <- function(s) {
    matches <- regmatches(
      s,
      gregexpr("<U\\+[0-9A-Fa-f]{4,6}>", s, perl = TRUE)
    )[[1]]

    if (length(matches) == 1 && matches[1] == "") return(s)

    for (m in matches) {
      hex <- sub("<U\\+([0-9A-Fa-f]{4,6})>", "\\1", m, perl = TRUE)
      replacement <- intToUtf8(strtoi(hex, base = 16))
      s <- sub(m, replacement, s, fixed = TRUE)
    }

    s
  }

  vapply(x, fix_one, character(1))
}

remove_accents <- function(x) {
  x <- decode_unicode_tags(x)
  x <- enc2utf8(x)
  stringi::stri_trans_general(x, "Latin-ASCII")
}

#Test
remove_accents("Ronald Araújo")
remove_accents("Ángel Di María")
remove_accents("Cristiano Ronaldo")

[1] "Ronald Araujo"

[1] "Angel Di Maria"

[1] "Cristiano Ronaldo"

Seems to work pretty well, lets apply it to both of our datasets and then compare again

In [12]:
player_names <- player_names %>%
    mutate(name = remove_accents(name))

worldcup_players <- worldcup_players %>%
    mutate(full_name = remove_accents(full_name))

length(setdiff(worldcup_players$full_name, player_names$name))
length(worldcup_players$full_name)
setdiff(worldcup_players$full_name, player_names$name)

[1] 691

[1] 2379

[1] "Bailey Wright"              "Cameron Devlin"            
  [3] "not applicable Alisson"     "not applicable Danilo"     
  [5] "not applicable Marquinhos"  "not applicable Casemiro"   
  [7] "not applicable Fred"        "not applicable Richarlison"
  [9] "not applicable Neymar"      "not applicable Raphinha"   
 [11] "not applicable Weverton"    "not applicable Fabinho"    
 [13] "not applicable Antony"      "not applicable Rodrygo"    
 [15] "not applicable Ederson"     "not applicable Bremer"     
 [17] "not applicable Pedro"       "Jerome Ngom Mbekeli"       
 [19] "Nicolas Nkoulou"            "Moumi Ngamaleu"            
 [21] "Georges-Kevin Nkoudou"      "Andre-Frank Zambo Anguissa"
 [23] "Eric Maxim Choupo-Moting"   "Samuel Gouet"              
 [25] "Nouhou Tolo"                "Souaibou Marou"            
 [27] "Liam Fraser"                "Lucas Cavallini"           
 [29] "Daniel Chacon"              "Keysher Fuller"            
 [31] "Johan Venegas"              "Gerson Torres"             
 [33] "Douglas Lopez"              "Jens Stryger Larsen"       
 [35] "William Pacho"              "Romario Ibarra"            
 [37] "Michael Estrada"            "Xavier Arreaga"            
 [39] "Sebas Mendez"               "Alexander Dominguez"       
 [41] "Djorkaeff Reasco"           "Lawrence Ati-Zigi"         
 [43] "Abdul Fatawu Issahaku"      "Ibrahim Danlad"            
 [45] "Abdul Manaf Nurudeen"       "Baba Rahman"               
 [47] "Shojae Khalilzadeh"         "Hossein Kanaanizadegan"    
 [49] "Rouzbeh Cheshmi"            "Mehdi Torabi"              
 [51] "Ahmad Nourollahi"           "Hossein Hosseini"          
 [53] "Abolfazl Jalali"            "Alfredo Talavera"          
 [55] "Munir Mohamedi"             "Ilias Chair"               
 [57] "Ahmed Reda Tagnaouti"       "Bilal El Khannous"         
 [59] "Badr Benoun"                "Yahya Jabrane"             
 [61] "not applicable Pepe"        "not applicable Vitinha"    
 [63] "not applicable Otavio"      "Saad Al-Sheeb"             
 [65] "not applicable Ro-Ro"       "Abdulaziz Hatem"           
 [67] "Ali Assadalla"              "Hassan Al-Haydos"          
 [69] "Musab Kheder"               "Homam Ahmed"               
 [71] "Ismaeel Mohammad"           "Khalid Muneer"             
 [73] "Salem Al-Hajri"             "Yousef Hassan"             
 [75] "Naif Al-Hadhrami"           "Mostafa Meshaal"           
 [77] "Mohammed Al-Rubaie"         "Ali Al-Bulaihi"            
 [79] "Mohammed Al-Breik"          "Abdulellah Al-Malki"       
 [81] "Hassan Al-Tambakti"         "Abdulrahman Al-Aboud"      
 [83] "Moussa N'Diaye"             "Moustapha Name"            
 [85] "Srdan Babic"                "Filip Duricic"             
 [87] "Yoon Jong-gyu"              "Kim Jin-su"                
 [89] "Kim Min-jae"                "Hwang In-beom"             
 [91] "Paik Seung-ho"              "Cho Gue-sung"              
 [93] "Song Bum-keun"              "Son Jun-ho"                
 [95] "Kim Moon-hwan"              "Hwang Ui-jo"               
 [97] "Na Sang-ho"                 "Lee Kang-in"               
 [99] "Kwon Kyung-won"             "Kwon Chang-hoon"           
[101] "Kim Tae-hwan"               "Cho Yu-min"                
[103] "Jeong Woo-yeong"            "Song Min-kyu"              
[105] "not applicable Koke"        "not applicable Gavi"       
[107] "not applicable Rodri"       "Dani Carvajal"             
[109] "not applicable Pedri"       "Aymen Mathlouthi"          
[111] "Bilel Ifa"                  "Hannibal Mejbri"           
[113] "Taha Yassine Khenissi"      "Ali Maaloul"               
[115] "Mohamed Ali Ben Romdhane"   "Ghailene Chaalali"         
[117] "Seifeddine Jaziri"          "Anis Ben Slimane"          
[119] "Jose Gimenez"               "Giorgian De Arrascaeta"    
[121] "Sebastian Sosa"             "Joe Morrell"               
[123] "Jonny Williams"             "James Meredith"            
[125] "Tim Cahill"                

Our set difference went down 165 observations! Lets keep trouble shooting.

There seems to be a few players with "not applicable" as their first name. Let's explore this issue.

In [13]:
worldcup_players %>%
    filter(grepl("not applicable", full_name)) %>%
        print(n=5)

worldcup_players %>%
    filter(grepl("not applicable", full_name)) %>%
        distinct(full_name) %>%
            nrow()

# A tibble: 88 x 3
  player_id full_name                 tournament_id
  <chr>     <chr>                     <chr>        
1 P-21531   not applicable Alisson    WC-2022      
2 P-40137   not applicable Danilo     WC-2022      
3 P-76060   not applicable Marquinhos WC-2022      
4 P-02305   not applicable Casemiro   WC-2022      
5 P-74941   not applicable Fred       WC-2022      
# i 83 more rows


[1] 85

In [14]:
readRDS(here("1.DataCleaning-R", "Data", "RDS", "FullRoster.rds")) %>%
    filter(given_name== "not applicable") %>%
        print(n=10)

# A tibble: 113 x 9
   tournament_id team_name team_id team_code player_id family_name   given_name 
   <chr>         <chr>     <chr>   <chr>     <chr>     <chr>         <chr>      
 1 WC-2010       Brazil    T-09    BRA       P-94551   "Maicon"      not applic~
 2 WC-2010       Brazil    T-09    BRA       P-42918   "L\u00facio"  not applic~
 3 WC-2010       Brazil    T-09    BRA       P-98607   "Juan"        not applic~
 4 WC-2010       Brazil    T-09    BRA       P-78264   "Elano"       not applic~
 5 WC-2010       Brazil    T-09    BRA       P-58388   "Kak\u00e1"   not applic~
 6 WC-2010       Brazil    T-09    BRA       P-13354   "Robinho"     not applic~
 7 WC-2010       Brazil    T-09    BRA       P-32968   "Luis\u00e3o" not applic~
 8 WC-2010       Brazil    T-09    BRA       P-30363   "Gilberto"    not applic~
 9 WC-2010       Brazil    T-09    BRA       P-74066   "Josu\u00e9"  not applic~
10 WC-2010       Brazil    T-09    BRA       P-86499   "Ramires"     not applic~
# i 103 

In [15]:
#Let's check some of these players on the Kaggle dataset
set.seed(2026)

player_names %>% 
    select(name) %>%
        filter(grepl("Maicon|Elano", name)) %>%
            slice_sample(n=10)

name
<chr>
Elano Yegen
Maicon
Maicon Oliveira
Maicon
Maicon


Some of the players seem to be only reffered to by one name in both datasets. Lets take out the not applicable and see improvements.

In [16]:
remove_not_applicable <- function(x) {
  gsub("not applicable\\s+", "", x)
}

worldcup_players <- worldcup_players %>%
  mutate(full_name = remove_not_applicable(full_name))

length(setdiff(worldcup_players$full_name, player_names$name))
length(worldcup_players$full_name)
setdiff(worldcup_players$full_name, player_names$name)

[1] 617

[1] 2379

[1] "Bailey Wright"              "Cameron Devlin"            
  [3] "Jerome Ngom Mbekeli"        "Nicolas Nkoulou"           
  [5] "Moumi Ngamaleu"             "Georges-Kevin Nkoudou"     
  [7] "Andre-Frank Zambo Anguissa" "Eric Maxim Choupo-Moting"  
  [9] "Samuel Gouet"               "Nouhou Tolo"               
 [11] "Souaibou Marou"             "Liam Fraser"               
 [13] "Lucas Cavallini"            "Daniel Chacon"             
 [15] "Keysher Fuller"             "Johan Venegas"             
 [17] "Gerson Torres"              "Douglas Lopez"             
 [19] "Jens Stryger Larsen"        "William Pacho"             
 [21] "Romario Ibarra"             "Michael Estrada"           
 [23] "Xavier Arreaga"             "Sebas Mendez"              
 [25] "Alexander Dominguez"        "Djorkaeff Reasco"          
 [27] "Lawrence Ati-Zigi"          "Abdul Fatawu Issahaku"     
 [29] "Ibrahim Danlad"             "Abdul Manaf Nurudeen"      
 [31] "Baba Rahman"                "Shojae Khalilzadeh"        
 [33] "Hossein Kanaanizadegan"     "Rouzbeh Cheshmi"           
 [35] "Mehdi Torabi"               "Ahmad Nourollahi"          
 [37] "Hossein Hosseini"           "Abolfazl Jalali"           
 [39] "Alfredo Talavera"           "Munir Mohamedi"            
 [41] "Ilias Chair"                "Ahmed Reda Tagnaouti"      
 [43] "Bilal El Khannous"          "Badr Benoun"               
 [45] "Yahya Jabrane"              "Saad Al-Sheeb"             
 [47] "Ro-Ro"                      "Abdulaziz Hatem"           
 [49] "Ali Assadalla"              "Hassan Al-Haydos"          
 [51] "Musab Kheder"               "Homam Ahmed"               
 [53] "Ismaeel Mohammad"           "Khalid Muneer"             
 [55] "Salem Al-Hajri"             "Yousef Hassan"             
 [57] "Naif Al-Hadhrami"           "Mostafa Meshaal"           
 [59] "Mohammed Al-Rubaie"         "Ali Al-Bulaihi"            
 [61] "Mohammed Al-Breik"          "Abdulellah Al-Malki"       
 [63] "Hassan Al-Tambakti"         "Abdulrahman Al-Aboud"      
 [65] "Moussa N'Diaye"             "Moustapha Name"            
 [67] "Srdan Babic"                "Filip Duricic"             
 [69] "Yoon Jong-gyu"              "Kim Jin-su"                
 [71] "Kim Min-jae"                "Hwang In-beom"             
 [73] "Paik Seung-ho"              "Cho Gue-sung"              
 [75] "Song Bum-keun"              "Son Jun-ho"                
 [77] "Kim Moon-hwan"              "Hwang Ui-jo"               
 [79] "Na Sang-ho"                 "Lee Kang-in"               
 [81] "Kwon Kyung-won"             "Kwon Chang-hoon"           
 [83] "Kim Tae-hwan"               "Cho Yu-min"                
 [85] "Jeong Woo-yeong"            "Song Min-kyu"              
 [87] "Dani Carvajal"              "Aymen Mathlouthi"          
 [89] "Bilel Ifa"                  "Hannibal Mejbri"           
 [91] "Taha Yassine Khenissi"      "Ali Maaloul"               
 [93] "Mohamed Ali Ben Romdhane"   "Ghailene Chaalali"         
 [95] "Seifeddine Jaziri"          "Anis Ben Slimane"          
 [97] "Jose Gimenez"               "Giorgian De Arrascaeta"    
 [99] "Sebastian Sosa"             "Joe Morrell"               
[101] "Jonny Williams"             "James Meredith"            
[103] "Tim Cahill"                 "Dimitri Petratos"          
[105] "Oscar Murillo"              "Mateus Uribe"              
[107] "Farid Diaz"                 "Jose Fernando Cuadrado"    
[109] "Jhonny Acosta"              "Daniel Colindres"          
[111] "Patrick Pemberton"          "David Guzman"              
[113] "Marco Urena"                "Kenner Gutierrez"          
[115] "Leonel Moreira"             "Essam El Hadary"           
[117] "Omar Gaber"                 "Ahmed Hegazi"              
[119] "Ahmed Fathy"                "Ayman Ashraf"              
[121] "Mohamed Abdel Shafy"        "Mahmoud Hamdy"             
[123] "Sherif Ekramy"              "Abdallah El Said"          
[125] "Saad Samir"                

Great! 74 out of the 85 players with not applicable were solved with this issue. Lets asses the other 11. We can do this by checking all full names with no spaces in the sett differences list. If we have 11 then we know we got all of the not applicables we're missing.

In [17]:
unmatched_players <- setdiff(worldcup_players$full_name, player_names$name)

one_word_unmatched <- unmatched_players[!grepl("\\s", unmatched_players)]

length(one_word_unmatched)
one_word_unmatched

[1] 11

[1] "Ro-Ro"     "Cedric"    "Elano"     "Kleberson" "Nilmar"    "Doni"     
 [7] "Grafite"   "Geremi"    "Yakubu"    "Miguel"    "Deco"

In [18]:
player_data_clean <- player_data %>%
  mutate(clean_name = remove_not_applicable(remove_accents(name)))

pattern <- paste0("\\b(", paste(one_word_unmatched, collapse = "|"), ")\\b")

player_data_clean %>%
  filter(grepl(pattern, clean_name)) %>%
    filter(!(clean_name %in% worldcup_players$full_name)) %>%
      distinct(name, clean_name, last_season, international_caps) %>%
        filter(last_season>= 2010 & international_caps > 0)



name,clean_name,last_season,international_caps
<chr>,<chr>,<int>,<int>
Miguel Lopes,Miguel Lopes,2024,4
Miguel Lloyd,Miguel Lloyd,2025,47
Luis Miguel Rodr<U+00ED>guez,Luis Miguel Rodriguez,2024,1
C<U+00E9>dric Soares,Cedric Soares,2025,34
C<U+00E9>dric Bakambu,Cedric Bakambu,2025,67
C<U+00E9>dric Hountondji,Cedric Hountondji,2024,29
Cedric Teuchert,Cedric Teuchert,2025,8
Pedro Miguel,Pedro Miguel,2025,99
Cedric Itten,Cedric Itten,2025,13


After further research Im comfortable assigning the full names to the following players.

In [19]:
worldcup_players <- worldcup_players %>%
  mutate(
    full_name = case_when(
      full_name == "Elano" ~ "Elano Yegen",
      full_name == "Cedric" ~ "Cedric Soares",
      full_name == "Ro-Ro" ~ "Pedro Miguel",
      TRUE ~ full_name
    )
  )

The other few players with one word names we will leave as undefined for now.


I will Have CODEX AI go thorugh the full list of differences and through the full p[layers.csv dataset and find clear minimal gramatical mismatches. I willl then manually check a few to minimize likelihood of errors.

In [20]:
# Obvious name fixes found by comparing unmatched World Cup names to players.csv
# Only includes high-confidence spelling, punctuation, or known single-name expansions.
worldcup_players <- worldcup_players %>%
  mutate(
    full_name = case_when(
      full_name == "Abdulrahman Al-Aboud" ~ "Abdulrahman Al-Oboud",
      full_name == "Eric Maxim Choupo-Moting" ~ "Eric-Maxim Choupo-Moting",
      full_name == "Lawrence Ati-Zigi" ~ "Lawrence Ati Zigi",
      full_name == "Abdulaziz Hatem" ~ "Abdulaziz Hatim",
      full_name == "Abdulellah Al-Malki" ~ "Abdulelah Al-Malki",
      full_name == "Ahmed Hegazi" ~ "Ahmed Hegazy",
      full_name == "Aleksandr Yerokhin" ~ "Aleksandr Erokhin",
      full_name == "Aleksei Ionov" ~ "Aleksey Ionov",
      full_name == "Aleksei Kozlov" ~ "Aleksey Kozlov",
      full_name == "Aleksei Miranchuk" ~ "Aleksey Miranchuk",
      full_name == "Andrey Yeshchenko" ~ "Andrey Eshchenko",
      full_name == "Antar Yahia" ~ "Anthar Yahia",
      full_name == "Bilal El Khannous" ~ "Bilal El Khannouss",
      full_name == "Brian Idowu" ~ "Bryan Idowu",
      full_name == "Cheick Tiote" ~ "Cheik Tiote",
      full_name == "Daler Kuzyayev" ~ "Daler Kuzyaev",
      full_name == "Dimitris Salpingidis" ~ "Dimitrios Salpingidis",
      full_name == "Dmitri Kombarov" ~ "Dmitriy Kombarov",
      full_name == "Erick Davis" ~ "Eric Davis",
      full_name == "Filip Duricic" ~ "Filip Djuricic",
      full_name == "Georgi Shchennikov" ~ "Georgiy Shchennikov",
      full_name == "Giorgian De Arrascaeta" ~ "Giorgian de Arrascaeta",
      full_name == "Guy N'dy Assembe" ~ "Guy Ndy Assembe",
      full_name == "Hannes THor Halldorsson" ~ "Hannes Thor Halldorsson",
      full_name == "Hassan Al-Haydos" ~ "Hasan Al-Haydos",
      full_name == "Marco Urena" ~ "Marcos Urena",
      full_name == "Mateus Uribe" ~ "Matheus Uribe",
      full_name == "Michael Babatunde" ~ "Michel Babatunde",
      full_name == "Sergei Ignashevich" ~ "Sergey Ignashevich",
      full_name == "Sotirios Kyrgiakos" ~ "Sotiris Kyrgiakos",
      full_name == "Srdan Babic" ~ "Srdjan Babic",
      full_name == "Steve Cherundolo" ~ "Steven Cherundolo",
      full_name == "Vasili Berezutski" ~ "Vasiliy Berezutski",
      full_name == "Vasilis Torosidis" ~ "Vasilios Torosidis",
      full_name == "Viktor Fayzulin" ~ "Viktor Faizulin",
      full_name == "William Pacho" ~ "Willian Pacho",
      full_name == "Yuri Lodygin" ~ "Yuriy Lodygin",
      full_name == "Yuri Zhirkov" ~ "Yuriy Zhirkov",
      full_name == "Andrey Lunyov" ~ "Andrey Lunev",
      full_name == "Dani Carvajal" ~ "Daniel Carvajal",
      full_name == "Fyodor Kudryashov" ~ "Fedor Kudryashov",
      full_name == "Fyodor Smolov" ~ "Fedor Smolov",
      full_name == "Georges-Kevin Nkoudou" ~ "Georges-Kevin N'Koudou",
      full_name == "Giorgos Karagounis" ~ "Georgios Karagounis",
      full_name == "Giorgos Tzavellas" ~ "Georgios Tzavellas",
      full_name == "John Heitinga" ~ "Johnny Heitinga",
      full_name == "Joo-ho Park" ~ "Ju-ho Park",
      full_name == "Mbark Boussoufa" ~ "Moubarak Boussoufa",
      full_name == "Nikos Spyropoulos" ~ "Nikolaos Spyropoulos",
      full_name == "Robert Green" ~ "Rob Green",
      full_name == "Yakubu" ~ "Yakubu Aiyegbeni",
      TRUE ~ full_name
    )
  )


In [21]:
length(setdiff(worldcup_players$full_name, player_names$name))
length(worldcup_players$full_name)
setdiff(worldcup_players$full_name, player_names$name)

[1] 563

[1] 2379

[1] "Bailey Wright"              "Cameron Devlin"            
  [3] "Jerome Ngom Mbekeli"        "Nicolas Nkoulou"           
  [5] "Moumi Ngamaleu"             "Andre-Frank Zambo Anguissa"
  [7] "Samuel Gouet"               "Nouhou Tolo"               
  [9] "Souaibou Marou"             "Liam Fraser"               
 [11] "Lucas Cavallini"            "Daniel Chacon"             
 [13] "Keysher Fuller"             "Johan Venegas"             
 [15] "Gerson Torres"              "Douglas Lopez"             
 [17] "Jens Stryger Larsen"        "Romario Ibarra"            
 [19] "Michael Estrada"            "Xavier Arreaga"            
 [21] "Sebas Mendez"               "Alexander Dominguez"       
 [23] "Djorkaeff Reasco"           "Abdul Fatawu Issahaku"     
 [25] "Ibrahim Danlad"             "Abdul Manaf Nurudeen"      
 [27] "Baba Rahman"                "Shojae Khalilzadeh"        
 [29] "Hossein Kanaanizadegan"     "Rouzbeh Cheshmi"           
 [31] "Mehdi Torabi"               "Ahmad Nourollahi"          
 [33] "Hossein Hosseini"           "Abolfazl Jalali"           
 [35] "Alfredo Talavera"           "Munir Mohamedi"            
 [37] "Ilias Chair"                "Ahmed Reda Tagnaouti"      
 [39] "Badr Benoun"                "Yahya Jabrane"             
 [41] "Saad Al-Sheeb"              "Ali Assadalla"             
 [43] "Musab Kheder"               "Homam Ahmed"               
 [45] "Ismaeel Mohammad"           "Khalid Muneer"             
 [47] "Salem Al-Hajri"             "Yousef Hassan"             
 [49] "Naif Al-Hadhrami"           "Mostafa Meshaal"           
 [51] "Mohammed Al-Rubaie"         "Ali Al-Bulaihi"            
 [53] "Mohammed Al-Breik"          "Hassan Al-Tambakti"        
 [55] "Moussa N'Diaye"             "Moustapha Name"            
 [57] "Yoon Jong-gyu"              "Kim Jin-su"                
 [59] "Kim Min-jae"                "Hwang In-beom"             
 [61] "Paik Seung-ho"              "Cho Gue-sung"              
 [63] "Song Bum-keun"              "Son Jun-ho"                
 [65] "Kim Moon-hwan"              "Hwang Ui-jo"               
 [67] "Na Sang-ho"                 "Lee Kang-in"               
 [69] "Kwon Kyung-won"             "Kwon Chang-hoon"           
 [71] "Kim Tae-hwan"               "Cho Yu-min"                
 [73] "Jeong Woo-yeong"            "Song Min-kyu"              
 [75] "Aymen Mathlouthi"           "Bilel Ifa"                 
 [77] "Hannibal Mejbri"            "Taha Yassine Khenissi"     
 [79] "Ali Maaloul"                "Mohamed Ali Ben Romdhane"  
 [81] "Ghailene Chaalali"          "Seifeddine Jaziri"         
 [83] "Anis Ben Slimane"           "Jose Gimenez"              
 [85] "Sebastian Sosa"             "Joe Morrell"               
 [87] "Jonny Williams"             "James Meredith"            
 [89] "Tim Cahill"                 "Dimitri Petratos"          
 [91] "Oscar Murillo"              "Farid Diaz"                
 [93] "Jose Fernando Cuadrado"     "Jhonny Acosta"             
 [95] "Daniel Colindres"           "Patrick Pemberton"         
 [97] "David Guzman"               "Kenner Gutierrez"          
 [99] "Leonel Moreira"             "Essam El Hadary"           
[101] "Omar Gaber"                 "Ahmed Fathy"               
[103] "Ayman Ashraf"               "Mohamed Abdel Shafy"       
[105] "Mahmoud Hamdy"              "Sherif Ekramy"             
[107] "Abdallah El Said"           "Saad Samir"                
[109] "Birkir Mar Saevarsson"      "Bjorn Bergmann Sigurdarson"
[111] "Holmar Orn Eyjolfsson"      "Hordur Bjorgvin Magnusson" 
[113] "Jon Dadi Bodvarsson"        "Mohammad Rashid Mazaheri"  
[115] "Mohammad Reza Khanzadeh"    "Pejman Montazeri"          
[117] "Tomoaki Makino"             "Jose de Jesus Corona"      
[119] "Hugo Ayala"                 "Rafael Marquez"            
[121] "Javier Hernandez"           "Jesus Manuel Corona"       
[123] "Oribe Peralta"              "Aziz Bouhaddouz"           
[125] "Mehdi Carcela"             

I want to assses the relevance of this unmatched players to our dataset and start considering the possibility of excluding some of these players.

In [22]:
nonmatched <- setdiff(worldcup_players$full_name, player_names$name)
fullworldcup_players <- readRDS(here("1.DataCleaning-R", "Data", "RDS", "FullRoster.rds"))

fullworldcup_players %>% 
    filter(full_name %in% nonmatched) %>%
        group_by(team_name, tournament_id) %>%
            summarize(players= sum(n())) %>%
                arrange(desc(players))




`summarise()` has grouped output by 'team_name'. You can override using the
`.groups` argument.


team_name,tournament_id,players
<chr>,<chr>,<int>
North Korea,WC-2010,23
New Zealand,WC-2010,18
South Korea,WC-2022,18
South Africa,WC-2010,17
Iran,WC-2014,15
South Korea,WC-2010,14
Australia,WC-2010,13
Japan,WC-2010,13
United States,WC-2010,12


Looking at this table we can see two clear trends. Firstly, some countries have almost all their players missing, there's a big chance that some of this countries leagues are simply not in our dataste and hence most of their national team players arent available. The other trend is that the oldest the tournamnet the most missing players. Since there is a good chance many of this players we just dont have data for I've decided to take all temas with 8 or more missing players out of our training data. 

In [23]:
fullworldcup_players %>% 
  filter(full_name %in% nonmatched) %>%
    group_by(team_name, tournament_id) %>%
        summarize(players = n(), .groups = "drop") %>%
            filter(players >= 8) %>%
                summarize(total_players = sum(players))

563-265

total_players
<int>
265


[1] 298

Now we only have 298 missing players, this is reasonable at around 10% of total players. It's inevitable to work with some incomplete data and assuming most of the missing plkayers are lower profile players we can assume that there effect on team performances would be lower so the model shouldn't be affected too strongly.

In [24]:
matched_players <- worldcup_players %>%
  rename(player_id_fj = player_id) %>%
  mutate(wc_year = as.integer(sub("WC-", "", tournament_id))) %>%
  arrange(full_name, desc(wc_year)) %>%
  distinct(full_name, .keep_all = TRUE) %>%
  inner_join(
    player_names %>%
      rename(
        player_id_kag = player_id,
        full_name = name
      ),
    by = "full_name"
  ) %>%
  filter(wc_year <= last_season) %>%
  distinct(full_name, .keep_all = TRUE) %>%
  select(full_name, player_id_fj, player_id_kag)

I added some players manually using other tools, the idea was to find the most relevant unmatched players.

In [25]:
manual_matches <- tibble::tribble(
  ~full_name,            ~player_id_fj, ~player_id_kag,
  "Saul",                "P-52837",      148928,
  "Thiago",              "P-92520",      60444,
  "Jose Maria Gimenez",  "P-65659",      250845,
  "Gareth Bale",         "P-63927",      39381,
  "Andres Iniesta",	     "P-56330",      7600,
  "David",               "P-49097",      7980,
  "Nemanja Vidic",       "P-68901",      19726,
  "Dani Alves",          "P-09441",      15951,
  "Javier Mascherano",   "P-77063",      19981,
  "Mikel Jon Obi",       "P-01164",      30739,
  "Javier Hernandez",    "P-36515",      50935,
  "Diego Godin",         "P-78440",      54928,	
  "Paulinho",            "P-54013",      57229,
  "Konstantino Manolas", "P-23816",	     104752
)

matched_players<- bind_rows(matched_players, manual_matches)

matched_players <- matched_players %>%
  filter(!(player_id_kag == 85771))

In [26]:
str(matched_players)
head(matched_players)
dim(matched_players)
any(duplicated(matched_players$player_id_fj))
any(duplicated(matched_players$player_id_kag))

tibble [1,728 x 3] (S3: tbl_df/tbl/data.frame)
 $ full_name    : chr [1:1728] "Aaron Lennon" "Aaron Long" "Aaron Mooy" "Aaron Ramsdale" ...
 $ player_id_fj : chr [1:1728] "P-64864" "P-75522" "P-95847" "P-62368" ...
 $ player_id_kag: num [1:1728] 14221 271072 123951 427568 50057 ...


full_name,player_id_fj,player_id_kag
<chr>,<chr>,<dbl>
Aaron Lennon,P-64864,14221
Aaron Long,P-75522,271072
Aaron Mooy,P-95847,123951
Aaron Ramsdale,P-62368,427568
Aaron Ramsey,P-86951,50057
Abde Ezzalzouli,P-21052,724520


[1] 1728    3

[1] FALSE

[1] FALSE

In [31]:
saveRDS(matched_players, here("1.DataCleaning-R", "Data", "RDS", "Matching.rds"))